users/src/main/scala/users/services/usermanagement/interpreters.scala

```diff
@@ -0,0 +1,238 @@
+
+  def generateId(): Future[Id] =
+    Future.successful(Id(UUID.randomUUID().toString))
+
+  def get(
+      id: Id
+  ): Future[Error Either User] =
+    for {
+      maybeUser ← userRepository.get(id)
+      result = maybeUser.toRight(Error.NotFound)
+    } yield result
+
+  def signUp(
+      userName: UserName,
+      emailAddress: EmailAddress,
+      password: Option[Password]
+  ): Future[Error Either User] =
+    (for {
+      maybeUser ← EitherT[Future, Error, Option[User]] {
+        userRepository.getByUserName(userName).map(Right.apply)
+      }
+      id ← EitherT[Future, Error, Id](generateId().map(Right.apply))
+      result ← EitherT.fromEither[Future] {
+        if (maybeUser.nonEmpty) Left(Error.AlreadyExists: Error)
+        else
+          Right(User(id, userName, emailAddress, password, OffsetDateTime.now()))
+      }
+    } yield result).value
+
+  def updateEmail(
+      id: Id,
+      emailAddress: EmailAddress
+  ): Future[Error Either User] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else Right(user.updateEmailAddress(emailAddress, OffsetDateTime.now()))
+      }
+      _ ← EitherT(save(result))
+    } yield result).value
+
+  def updatePassword(
+      id: Id,
+      password: Password
+  ): Future[Error Either User] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else Right(user.updatePassword(password, OffsetDateTime.now()))
+      }
+      _ ← EitherT(save(result))
+    } yield result).value
+
+  def resetPassword(
+      id: Id
+  ): Future[Error Either User] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else Right(user.resetPassword(OffsetDateTime.now()))
+      }
+      _ ← EitherT(save(result))
+    } yield result).value
+
+  def block(
+      id: Id
+  ): Future[Error Either User] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else if (user.isBlocked) Left(Error.Blocked)
+        else Right(user.block(OffsetDateTime.now))
+      }
+      _ ← EitherT(save(result))
+    } yield result).value
+
+  def unblock(
+      id: Id
+  ): Future[Error Either User] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else if (user.isActive) Left(Error.Active)
+        else Right(user.unblock(OffsetDateTime.now))
+      }
+      _ ← EitherT(save(result))
+    } yield result).value
+
+  def delete(
+      id: Id
+  ): Future[Error Either Done] =
+    (for {
+      user ← EitherT(get(id))
+      result ← EitherT.fromEither[Future] {
+        if (user.isDeleted) Left(Error.Deleted)
+        else if (user.isActive) Left(Error.Active)
```

Comment:
delete only after block?

I.e. should user be able to delete their own accounts? or is this a discussion point?

Reply:
again, something they can work with if they want to.